### ISSUE TITLE
Cancellable SocketAsyncEventArgs methods

### ISSUE THREAD

**Issue by scalablecory** [dotnet/runtime#48477]
## Background and Motivation

We already have (or [will likely have](https://github.com/dotnet/runtime/issues/33418), for UDP) a cancellable implementation of these methods, but they are not public. I propose making them public.

There are some scenarios where using SocketAsyncEventArgs will be more performant than the Task-based methods, but you must currently trade off cancellation support or get at the private methods via reflection. I'd like to not depend on reflection anymore.

## Proposed API

```c#
class Socket
{
    // existing.
    public bool SendAsync(SocketAsyncEventArgs e);
    public bool ReceiveAsync(SocketAsyncEventArgs e);
    public bool SendToAsync(SocketAsyncEventArgs e);
    public bool SendPacketsAsync(SocketAsyncEventArgs e);
    public bool ReceiveFromAsync(SocketAsyncEventArgs e);
    public bool ReceiveMessageFromAsync(SocketAsyncEventArgs e);

    // new.
    public bool SendAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken);
    public bool ReceiveAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken);
    public bool SendToAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken);
    public bool SendPacketsAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken);
    public bool ReceiveFromAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken);
    public bool ReceiveMessageFromAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken);
}
```

CC @antonfirsov @geoffkizer

**Comment by antonfirsov** [comment]
This is a weird API that deviates from the typical async patterns, has a very strong hack-taste, and is unprecedented  in BCL. I'm not sure if exposing stuff like this is a good tradeoff.

Can't we instead introduce optimized `ValueTask` overloads for gathered reads/writes?

**Comment by scalablecory** [comment]
> This is a weird API that deviates from typical async patterns, has a very strong hack-taste, and is unprecedented in the BCL. I'm not sure if exposing stuff like this is a good tradeoff.

I disagree re: API being unprecedented-- `CancellationToken` is the standard model of cancellation; if anything it makes them less weird to add support.

Are you proposing that `SocketAsyncEventArgs` is dead and should be kept in maintenance mode going forward? I'd be glad to see that, but we should open an issue to track migrating the features it exposes to the `ValueTask` overloads, and testing that those have equivalent perf.

**Comment by antonfirsov** [comment]
> re: API being unprecedented -- `CancellationToken` is the standard model of cancellation

I'm talking about specific API models. `CancellationToken` is the standard, preferred way -- on `Task`-based methods. In your proposal you are mixing the EAP model with the Task-based model. I'm not aware of existing similar API-s in the BCL.

> Are you proposing that `SocketAsyncEventArgs` is dead and should be kept in maintenance mode going forward?

No, but it has different patterns for cancellation (eg. [`CancelConnectAsync(e)`](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.cancelconnectasync?view=net-5.0) which conforms to [this guideline](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/event-based-asynchronous-pattern-overview)). We should either follow that pattern and add `Socket.CancelReadAsync(e)` + `CancelWriteAsync(e)`, or add new `ValueTask`-based overloads, but not mix the two, even though it looks cheap to implement. 

My understanding is that we are committed to prefer Task-based stuff over EAP, when it comes to adding new API-s. If I'm not missing anything, the following addition should be enough to support the performance requirements of the "Flexible HTTP" stuff:

```C#
public class Socket
{
    public ValueTask WriteAsync(IReadOnlyList<ReadOnlyMemory<byte>> buffers, CancellationToken cancellationToken = default);
    public ValueTask<int> ReadAsync(IReadOnlyList<Memory<byte>> buffers, CancellationToken cancellationToken= default);
}
```

**Comment by geoffkizer** [comment]
> Are you proposing that SocketAsyncEventArgs is dead and should be kept in maintenance mode going forward? I'd be glad to see that, but we should open an issue to track migrating the features it exposes to the ValueTask overloads, and testing that those have equivalent perf.

"Dead" is probably a little too strong.

I think the Task based APIs are our primary and preferred API, and our long-term goal should be to ensure that users never need to use SAEA instead of Task APIs due to missing features or perf. And yeah, that means we should (a) track issues re missing functionality in the Task APIs vs SAEA. and (b) do perf evaluation to ensure there's no perf gap here.

The main gap I'm aware of today is scatter/gather support, but that's sort of tracked via the issue of whether/how to do scatter/gather in Stream and derived classes.

**Comment by antonfirsov** [comment]
> The main gap I'm aware of today is scatter/gather support, but that's sort of tracked via the issue of whether/how to do scatter/gather in Stream and derived classes.

I assume you mean #25344. That does not include the missing async socket API-s, and AFAIK there is no tracking issue for scatter/gather on `Socket` (two methods from https://github.com/dotnet/runtime/issues/48477#issuecomment-784246675).

**Comment by geoffkizer** [comment]
> I assume you mean #25344.

Yeah, that's what I meant.

**Comment by karelz** [comment]
Triage:
- Ideally move from SAEA to Task APIs long-term
- We need to commit to not just Sockets, but also on Stream (not implement, but commit)
- OK with perf concerns, but would be better to have real test comparing ValueTask and SAEA
- We need API proposal

**Comment by karelz** [comment]
Idea: Create replacement issue to better capture the final intent.

**Comment by dhhoang** [comment]
Hi, is this issue still active?  
I'd love to have `Socket.CancelReadAsync(e)/CancelWriteAsync(e)` as I think it will be inline with the design of `SocketAsyncEventArgs`. In which case the `SocketError` code would be set to a specific code e.g. `Interrupted`.  
My use case for this is to time-out individual socket calls. Currently that can be done with the `socket.SendAsync(data, CancellationToken)` overload but it means there is a cancellation registration for every socket call which would likely reduce performance.  

https://github.com/dotnet/runtime/blob/1e3544ec04e27538f3d38d274f7a65277f0d681f/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs#L195

**Comment by dhhoang** [comment]
> The overloads proposed here would not get rid of that CT registration, since we would be still passing the token. AFAIK most if not all CT based IO operations in .NET create a registration internally, but there several optimizations under the hood to make it as cheap as possible. Eg. in socket PAL in the cases where the socket operation completes synchronously, `ProcessIOCPResult` won't be invoked and no registration happens.

I am not proposing overloads with `CancellationToken` as I don't think it makes sense to have them in `SocketAsyncEventArgs`. I think the `SocketAsyncEventArgs` methods are supposed to give us lower-level control on socket operations e.g. getting a `SocketError` . What I am proposing are `Socket.CancelXAsync(e)` (X might be send/recv/ etc.).   
The underlying implementation would be to cancel the operation at the native level, for example, calling `CancelIoEx` on Windows, and on Linux I think it's about ignoring the `epoll` event.  Essentially it'd do the same action that's invoked when `CancellationToken` is triggered.

> Is it an assumption, or do you have data proving this?  

I don't have any data, and I'd be happy to create a benchmark to test this. My thinking is that `SendAsync(CancellationToken)` also uses SAEA, **plus** the CT registration, it would likely not be as efficient as using SAEA alone. For our use case we also expect a lot of pending read()/write().

**Comment by antonfirsov** [comment]
The standard way to achieve this would be to `Close`/`Dispose` the socket, which would cancel all pending IO. Of course this would terminate the TCP connection, but for most use-cases we've seen this is acceptable or even desirable. Do you still need to be able receive from the socket after cancelling sends or vice-versa? Note that usually it's not safe to eg. issue a new read after a cancelled one from the application POV, since the status of the cancelled operation is undefined (buffer sent/partially sent/not sent).

**Comment by dhhoang** [comment]
I've managed to create a benchmark. This is just a basic test, a simplified echo server: https://github.com/dhhoang/dotnet-socket-cancellation  
Please have a look at the readme for some results. I think it shows `CancellationToken` methods are noticeably more expensive that SAEA.
> Do you still need to be able receive from the socket after cancelling sends or vice-versa

Yes, that is indeed our use case :)

**Comment by Ricman666** [comment]
I also tend to make these overloads public visible. 

And there could also be overloads for synchronous methods with cancellable function too.
for example:

```
public static Socket Accept(this Socket obj, int timeout, CancellationToken stoppingToken);
public static void Connect(this Socket obj, EndPoint remoteEP, int timeout, CancellationToken stoppingToken);
public static void Disconnect(this Socket obj, bool reuseSocket, int timeout, CancellationToken stoppingToken);
public static int Receive(this Socket obj, byte[] buffer, int offset, int size, SocketFlags socketFlags, int timeout, CancellationToken stoppingToken);
public static int ReceiveFrom(this Socket obj, byte[] buffer, int offset, int size, SocketFlags socketFlags, ref EndPoint remoteEP, int timeout, CancellationToken stoppingToken);
public static int Send(this Socket obj, byte[] buffer, int offset, int size, SocketFlags socketFlags, int timeout, CancellationToken stoppingToken);
public static int SendTo(this Socket obj, byte[] buffer, int offset, int size, SocketFlags socketFlags, EndPoint remoteEP, int timeout, CancellationToken stoppingToken);
```
 
This is needed if you are writing a high performance server and have limited resources or need to quickly terminate the worker thread for the server. 

Task based methods have unpredictable behavior for these cases and cannot even limit resources to a single thread.
